pkg/adaptor/cloud/cloud.go

```diff
@@ -34,7 +34,7 @@ const (
 var logger = log.New(log.Writer(), "[adaptor/cloud] ", log.LstdFlags|log.Lmsgprefix)
 
 type Provider interface {
-	CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator) (instance *Instance, err error)
+	CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator, instanceType string) (instance *Instance, err error)
```

Comment:
I think it's OK for the first implementation, but I think we need a mechanism to select an instance type based on required vcpus and memory size.

How about adding a new method like `SelectInstanceType` in the `Provider` interface, instead of adding a new argument to `CreateInstance`?

```
type Provider interface {
	CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator) (instance *Instance, err error)
	DeleteInstance(ctx context.Context, instanceID string) error
	Teardown() error
	SelectInstanceType(vcpus, memory int) string
}
```

Reply:
Few thoughts on using vcpus and memory size to decide on the instance type:
1. If we want to use cpu/mem request/limit data from resource attribute in the pod spec then we need to update the webhook to not remove resource attribute from the spec.  
2. If we want to use cpu/mem annotation, then we will need to pass the required annotations to CAA and figure out if we need to reconcile with any cpu/mem resource request/limit in the spec. Most likely we can have the same logic as with Kata local hypervisor but without the hotplugging stuff.
3. We'll need to work out a best fit function which will return the closest instance type match from the configured list based on the cpu/mem request. This was something which @mkulke had suggested initially

@yoheiueda I'm ok to rewrite this PR and incorporate the additional interface if it avoids immediate rewrite later on. Let me know wdyt?

@mkulke @snir911 @jtumber-ibm since you have acked this PR, wanted to hear your thoughts as well.

Reply:
I think both the approaches will be needed - instance type and cpu/mem will be helpful to the end user. Personally I find it hard to remember the cpu/mem of the instance types.

Reply:
I wonder whether we can retrieve the vCPU + memory characteristics from provider APIs, so when we process the ConfigMap initially, the provided instances types are enriched with this data. So we don't need the user to provide this, and we can sort out invalid instance types.

I wouldn't go beyond some verification check like "will pod fit into the given instance type" and minimal logic "if no instance type is given, pick the smallest possible from the configured available instance types" (that decision will probably be a bit arbitrary).

Reply:
@mkulke it's possible to get the vcpu and memory characteristics.
Also I was discussing with @yoheiueda on updating the CreateInstance method signature. It'll be needed since the provider apis need the instance type for VM creation. 

I'll update this PR to also include vcpu/mem annotation handling for instance type selection which we can keep improving going forward

Reply:
Could we also add architecture into the specifications for choosing a instance type? Do we then also need to include confidential VM selection too? Also, gpus might be important use-case in the future?

Reply:
You can use the instance type annotation to select gpu or any other instance without using cpu/mem